### DISCUSSION TITLE
Surrounding multiple lines with the same tag

### DISCUSSION THREAD

**Discussion by jasonlong** [nvim-mini/mini.nvim#763] [General]
Apologies if this has already been answered, but I didn't find anything matching exactly this after some searching here.

I know the common way of doing this is with macros, but I'm curious if there's a "mini" way.

Say you have three lines of varying length:

```
first item
second item
third item
```

and you want to do two things:
1. Wrap the whole block in a `<ul>` tag
2. Wrap each item in a `<li>` tag

The first one is easy enough with `sat`, but I've never landed on a nice way to do the second.

The final result would be

```html
<ul>
  <li>first item</li>
  <li>second item</li>
  <li>third item</li>
</ul>

**Comment by echasnovski** [comment]
Hi, Jason! Thanks for using 'mini.nvim'! It means a lot!

> I know the common way of doing this is with macros, but I'm curious if there's a "mini" way.

Well, I'd say that the best approach in this particular case is to use dot-repeat and `_` textobject (which represents "line"). That is:
- Put cursor on any column of "first item" line.
- Press `sa_t` followed by `li` and `<CR>`. It should transform the line into desired one.
- Press `j.` until completed.

------

**But**, judging by "The first one is easy enough with `sat`", you prefer `respect_selection_type = true`. This makes things a bit not straightforward. Mostly because there are conflicting use cases: for `<ul>` it would be better to respect selection type, but for `<li>` - not. The second part would require having a textobject which represents line but is itself charwise (not linewise).

But fear not, as there is a "mini" way! There are ['mini.ai'](https://github.com/echasnovski/mini.nvim/blob/main/readmes/mini-ai.md) to create textobjects and [`gen_ai_spec.line()` in 'mini.extra'](https://github.com/echasnovski/mini.nvim/blob/a683bfe8e03293e3bb079e24c94e51977756a3ec/doc/mini-extra.txt#L150) which can achieve exactly this. So the use case can be resolved with something like this:
- Set up 'mini.ai' module with custom textobject:
    ```lua
    require('mini.ai').setup({
      custom_textobjects = {
        L = require('mini.extra').gen_ai_spec.line(),
      }
    })
    ```
- Follow steps from the first approach but use `iL` textobject instead of `_`.

(Full disclosure: I hoped that "forcing" charwise mode with `sav_t` (notice `v`) would work, but apparently it is not really compatible with `_` textobject)

------

I'd personally would go the first route. This makes wrapping paragraph with `<ul>` not as straightforward, but as this is a usually a less common task, should be fine.

**Reply by jasonlong** [reply]
Wow thanks for the great explanation, @echasnovski! I can't believe I wasn't familiar w/ the `_` text object. That changes everything. That first method seems great to me, but I will also dig into `mini.ai` and `mini.extra`. 

Keep up the great work on this project! 🤘🏼
